### PR TITLE
Amazon kms_facts module

### PR DESCRIPTION
**Pull request by willthames**
##### SUMMARY
Add a module for KMS facts

##### ISSUE TYPE
 - New Module Pull Request

##### COMPONENT NAME
kms_facts

##### ANSIBLE VERSION
```
ansible 2.4.0 (devel 0b784c65b1) last updated 2017/07/13 15:27:10 (GMT +1000)
  config file = /etc/ansible/ansible.cfg
  configured module search path = [u'/home/will/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /home/will/src/ansible/lib/ansible
  executable location = /home/will/src/ansible/bin/ansible
  python version = 2.7.13 (default, May 10 2017, 20:04:28) [GCC 6.3.1 20161221 (Red Hat 6.3.1-1)]
```

### CODE REVIEW

File: lib/ansible/modules/cloud/amazon/kms_facts.py [hunk 1]
```diff
@@ -0,0 +1,313 @@
...
+description:
+    - Gather facts about AWS KMS keys including tags and grants
+version_added: "2.4"
+author: "Will Thames (@willthames)"
+options:
+  filters:
+    description:
+      - A dict of filters to apply. Each dict item consists of a filter key and a filter value.
+        The filters aren't natively supported by boto3, but are supported to provide similar
+        functionality to other modules. Standard tag filters (C(tag-key), C(tag-value) and
```
Comment (s-hertel) [inline lib/ansible/modules/cloud/amazon/kms_facts.py]: Is C() only for files?
http://docs.ansible.com/ansible/dev_guide/developing_modules_documenting.html#formatting-options

Comment (s-hertel) [inline lib/ansible/modules/cloud/amazon/kms_facts.py]: Is C() only for files?
http://docs.ansible.com/ansible/dev_guide/developing_modules_documenting.html#formatting-options

File: lib/ansible/modules/cloud/amazon/kms_facts.py [hunk 2]
```diff
@@ -0,0 +1,313 @@
...
+    # Handle pagination here as list_resource_tags does not have
+    # a paginator
+    kwargs = {}
+    tags = []
+    more = True
+    while more:
+        try:
+            tag_response = get_kms_tags_with_backoff(connection, key_id, **kwargs)
+            tags.extend(tag_response['Tags'])
+        except botocore.exceptions.ClientError as e:
```
Comment (s-hertel) [inline lib/ansible/modules/cloud/amazon/kms_facts.py]: It looks like the error response for keys that don't have the option of having tags and insufficient permissions for getting tags are the same. Maybe we shouldn't fail here if e.response['Error']['Code'] == 'AccessDeniedException' and just let the keys be empty. I'm not sure if there's a way of differentiating between the permissions and the acceptable key options for special keys.

File: lib/ansible/modules/cloud/amazon/kms_facts.py [hunk 3]
```diff
@@ -0,0 +1,313 @@
...
+            module.fail_json(msg="Failed to obtain key tags",
+                             exception=traceback.format_exc(),
+                             **camel_dict_to_snake_dict(e.response))
+        if tag_response.get('NextMarker'):
+            kwargs['Marker'] = tag_response['NextMarker']
+        else:
+            more = False
+    return boto3_tag_list_to_ansible_dict(tags)
```
Comment (s-hertel) [inline lib/ansible/modules/cloud/amazon/kms_facts.py]: I think this should be `return boto3_tag_list_to_ansible_dict(tags, 'TagKey', 'TagValue')` to get any tags back
http://boto3.readthedocs.io/en/latest/reference/services/kms.html#list_resource_tags

File: lib/ansible/modules/cloud/amazon/kms_facts.py [hunk 4]
```diff
@@ -0,0 +1,313 @@
...
+    region, ec2_url, aws_connect_params = get_aws_connection_info(module, boto3=True)
+
+    if region:
+        connection = boto3_conn(module, conn_type='client', resource='kms', region=region, endpoint=ec2_url, **aws_connect_params)
+    else:
+        module.fail_json(msg="region must be specified")
+
+    all_keys = get_kms_facts(connection, module)
+    filtered = [key for key in all_keys if key_matches_filters(key, module.params['filters'])]
+    module.exit_json(keys=[camel_dict_to_snake_dict(key) for key in filtered])
```
Comment (s-hertel) [inline lib/ansible/modules/cloud/amazon/kms_facts.py]: The tag keys are being lowercased (maybe not a big deal, but I spent 10 minutes trying to figure out why filtering by them wasn't working).

Comment (willthames) [reply lib/ansible/modules/cloud/amazon/kms_facts.py]: You've done more testing on the tag functionality than I have, sorry!

It's always best to do camel_to_snake before doing tag conversion, I should know this by now.

File: lib/ansible/modules/cloud/amazon/kms_facts.py (hunk 1 above)
Comment (willthames) [reply lib/ansible/modules/cloud/amazon/kms_facts.py]: `C()` is for option values. Here `filters` is the key so `tag-key` is part of the value (but also part of a key in the value). It's arguable, I'd say.

Comment (s-hertel) [reply lib/ansible/modules/cloud/amazon/kms_facts.py]: I apparently cannot read.

File: lib/ansible/modules/cloud/amazon/aws_kms_facts.py [hunk 5]
```diff
@@ -0,0 +1,355 @@
...
+@AWSRetry.backoff(tries=5, delay=5, backoff=2.0)
+def get_kms_keys_with_backoff(connection):
+    paginator = connection.get_paginator('list_keys')
+    return paginator.paginate().build_full_result()
+
+
+@AWSRetry.backoff(tries=5, delay=5, backoff=2.0)
+def get_kms_aliases_with_backoff(connection):
+    paginator = connection.get_paginator('list_aliases')
+    return paginator.paginate().build_full_result()['Aliases']
```
Comment (ryansb) [inline lib/ansible/modules/cloud/amazon/aws_kms_facts.py]: The other `get_kms_*` functions return the full response. Can you either return the full response here, or standardize on only returning the objects?

**Review by s-hertel — APPROVED** [review]

**Comment by s-hertel** [comment]
shipit

**Review by ryansb — CHANGES_REQUESTED** [review]

**Comment by willthames** [comment]
!needs_revision

**Comment by s-hertel** [comment]
I think rebasing resets the ansibot's stale_review label process, and then it takes another [week](https://github.com/ansible/ansibullbot/blob/c665e5b0e00bff15bd68457883baccdfefd60a7e/ansibullbot/triagers/plugins/needs_revision.py#L359) to remove needs_revision and add stale_review.

**Comment by willthames** [comment]
@jctanner mentioned that the bot won't remove needs_revision until @ryansb's rejection has been removed.

**Review by ryansb — APPROVED** [review]

**Comment by s-hertel** [comment]
bot_status

**Pull request merged**